[GRPC] Documentation for 3.2 GRPC GA (Bulk and k-NN Search)

karenyrx opened the pull request:
### Description
**Main changes**
- `index.md`:
     - Remove mentions of 'plugin' terminology (as the transport-grpc has been moved to a module)
     - Update gRPC installation instructions and advanced configuration settings
     - Callout Bulk and KNN APIs as generally available 
- `knn.md`:
    - Add new k-NN gRPC API documentation (and add a link to it from the HTTP k-NN page). 
    - Mark KNN as generally available (with limited support for `filter`)
- `bulk.md` / `search.md` :
    - API schema changes, resulting from upgrading opensearch-protobufs from 0.3.0 -> 0.6.0. 
    - Mark Bulk API as generally available 
- Vale configuration:
    - Add `gRPC` to vocabulary exceptions

### Issues Resolved
Closes #10556 

### Version
_List the OpenSearch version to which this PR applies, e.g. 2.14, 2.12--2.14, or all._
3.2 

### Frontend features
_If you're submitting documentation for an OpenSearch Dashboards feature, add a video that shows how a user will interact with the UI step by step. A voiceover is optional._ 

### Checklist
- [x] By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license and subject to the [Developers Certificate of Origin](https://github.com/opensearch-project/OpenSearch/blob/main/CONTRIBUTING.md#developer-certificate-of-origin).
For more information on following Developer Certificate of Origin and signing off your commits, please check [here](https://github.com/opensearch-project/OpenSearch/blob/main/CONTRIBUTING.md#developer-certificate-of-origin).

kolchfa-aws commented on _api-reference/grpc-apis/index.md:
```diff
@@ -50,29 +63,86 @@ To enable the gRPC plugin (`transport-grpc`) in OpenSearch:
+
+### Notes
+- For duration-based settings (e.g., `max_connection_age`), you can use units such as `ms` (milliseconds), `s` (seconds), `m` (minutes).
+- For size-based settings (e.g., `max_msg_size`), you can use units such as `b` (bytes), `kb`, `mb`, `gb`.
```
```suggestion
```

kolchfa-aws commented on _api-reference/grpc-apis/index.md:
```diff
@@ -50,29 +63,86 @@ To enable the gRPC plugin (`transport-grpc`) in OpenSearch:
+### Notes
+- For duration-based settings (e.g., `max_connection_age`), you can use units such as `ms` (milliseconds), `s` (seconds), `m` (minutes).
+- For size-based settings (e.g., `max_msg_size`), you can use units such as `b` (bytes), `kb`, `mb`, `gb`.
+- All settings are node-scoped unless otherwise specified.
```
```suggestion
All settings are node-scoped unless otherwise specified.
{: .note}
```

kolchfa-aws commented on _api-reference/grpc-apis/index.md:
```diff
@@ -50,29 +63,86 @@ To enable the gRPC plugin (`transport-grpc`) in OpenSearch:
+- For size-based settings (e.g., `max_msg_size`), you can use units such as `b` (bytes), `kb`, `mb`, `gb`.
+- All settings are node-scoped unless otherwise specified.
+
+These settings are similar to the [HTTP Network settings]({{site.url}}{{site.baseurl}}/install-and-configure/configuring-opensearch/network-settings/#advanced-http-settings) but specifically apply to gRPC communication. For more details about the `transport-grpc `module implementation, see the [OpenSearch `transport-grpc` module](https://github.com/karenyrx/OpenSearch/tree/protobufs-0.8.0-transpo ...
```
```suggestion
These settings are similar to the [HTTP network settings]({{site.url}}{{site.baseurl}}/install-and-configure/configuring-opensearch/network-settings/#advanced-http-settings) but specifically apply to gRPC communication. For more information about the `transport-grpc `module implementation, see the [OpenSearch `transport-grpc` module](https://github.com/karenyrx/OpenSearch/tree/protobufs-0.8.0-transport-grpc-fixes/modules/transport-grpc).
```

karenyrx replied:
Removed

kolchfa-aws commented on _api-reference/grpc-apis/knn.md:
```diff
@@ -0,0 +1,148 @@
+**Introduced 3.2**
+{: .label .label-purple }
+
+**Generally available 3.2**
```
```suggestion
```

kolchfa-aws commented on _api-reference/grpc-apis/knn.md:
```diff
@@ -0,0 +1,148 @@
+{: .label .label-purple }
+
+**Generally available 3.2**
+{: .label .label-green }
```
```suggestion
```

kolchfa-aws commented on _api-reference/grpc-apis/knn.md:
```diff
@@ -0,0 +1,148 @@
+**Generally available 3.2**
+{: .label .label-green }
+
+The gRPC k-NN API is generally available starting with OpenSearch 3.2. However, expect tweaks to the protobuf structure as the feature matures in the next releases.
```
```suggestion
The gRPC k-NN API is generally available starting with OpenSearch 3.2. However, expect updates to the protobuf structure as the feature matures in the next releases.
```

kolchfa-aws commented on _api-reference/grpc-apis/knn.md:
```diff
@@ -0,0 +1,148 @@
+
+The gRPC k-NN API is generally available starting with OpenSearch 3.2. However, expect tweaks to the protobuf structure as the feature matures in the next releases.
+
+The gRPC k-NN API provides an efficient, binary-encoded interface for performing k-nearest neighbor searches using protocol buffers over gRPC. k-NN (k-nearest neighbor) is a plugin that offers a specific search query type for vector similarity searches. This API offers superior performance compared to the traditional HTTP-based approach, making it ideal for large-scale machine learning and vector ...
```
```suggestion
The gRPC k-NN API provides an efficient, binary-encoded interface for performing k-nearest neighbor searches using protocol buffers over gRPC. The k-NN plugin offers a specific search query type for vector similarity searches. This API offers superior performance compared to the traditional HTTP-based approach, making it ideal for large-scale machine learning and vector database applications.
```

kolchfa-aws commented on _api-reference/grpc-apis/knn.md:
```diff
@@ -0,0 +1,148 @@
+
+gRPC k-NN APIs reside in the [`SearchService`](https://github.com/opensearch-project/opensearch-protobufs/blob/0.6.0/protos/services/search_service.proto#L22), the same service used for general search operations.
+
+You can submit k-NN search requests by invoking the [`Search`](https://github.com/opensearch-project/opensearch-protobufs/blob/0.6.0/protos/services/search_service.proto#L23) gRPC method within the `SearchService`, using a [`KnnQuery`](https://github.com/opensearch-project/opensearch-protobufs/blob/0.6.0/protos/schemas/common.proto#L1058) within the search request. The method takes in a [`SearchR ...
```
```suggestion
You can submit k-NN search requests by invoking the [`Search`](https://github.com/opensearch-project/opensearch-protobufs/blob/0.6.0/protos/services/search_service.proto#L23) gRPC method within the `SearchService`, using a [`KnnQuery`](https://github.com/opensearch-project/opensearch-protobufs/blob/0.6.0/protos/schemas/common.proto#L1058) within the search request. The method takes a [`SearchRequest`](https://github.com/opensearch-project/opensearch-protobufs/blob/0.6.0/protos/schemas/search.proto#L18) and returns a [`SearchResponse`](https://github.com/opensearch-project/opensearch-protobufs/blob/0.6.0/protos/schemas/search.proto#L317).
```

kolchfa-aws commented on _api-reference/grpc-apis/knn.md:
```diff
@@ -0,0 +1,148 @@
+
+You can submit k-NN search requests by invoking the [`Search`](https://github.com/opensearch-project/opensearch-protobufs/blob/0.6.0/protos/services/search_service.proto#L23) gRPC method within the `SearchService`, using a [`KnnQuery`](https://github.com/opensearch-project/opensearch-protobufs/blob/0.6.0/protos/schemas/common.proto#L1058) within the search request. The method takes in a [`SearchR ...
+
+The gRPC implementation leverages the same underlying k-NN functionality as the HTTP API while providing improved performance characteristics through protocol buffer serialization.
```
```suggestion
The gRPC implementation uses the same underlying k-NN functionality as the HTTP API while providing improved performance through protocol buffer serialization.
```

kolchfa-aws commented on _api-reference/grpc-apis/knn.md:
```diff
@@ -0,0 +1,148 @@
+
+## KnnQuery fields
+
+The gRPC k-NN API uses the [`KnnQuery`](https://github.com/opensearch-project/opensearch-protobufs/blob/0.6.0/protos/schemas/common.proto#L1058) message within a [`QueryContainer`](https://github.com/opensearch-project/opensearch-protobufs/blob/0.6.0/protos/schemas/common.proto#L341) for k-nearest neighbor searches. The `KnnQuery` message accepts the following fields:
```
```suggestion
The gRPC k-NN API uses the [`KnnQuery`](https://github.com/opensearch-project/opensearch-protobufs/blob/0.6.0/protos/schemas/common.proto#L1058) message within a [`QueryContainer`](https://github.com/opensearch-project/opensearch-protobufs/blob/0.6.0/protos/schemas/common.proto#L341) for k-NN searches. The `KnnQuery` message accepts the following fields.
```

kolchfa-aws commented on _api-reference/grpc-apis/knn.md:
```diff
@@ -0,0 +1,148 @@
+| Field | Protobuf type | Description |
+| :---- | :---- | :---- |
+| `field` | `string` | The vector field against which to run the search query. Required. |
+| `vector` | `repeated float` | Query vector. Must have the same number of dimensions as the vector field. Optional. |
```
```suggestion
| `vector` | `repeated float` | A query vector. Must have the same number of dimensions as the vector field. Optional. |
```

kolchfa-aws commented on _api-reference/grpc-apis/knn.md:
```diff
@@ -0,0 +1,148 @@
+| `min_score` | `float` | The minimum similarity score for a neighbor to be considered a hit. Optional. |
+| `max_distance` | `float` | The maximum physical distance in vector space for a neighbor to be considered a hit. Optional. |
+| `filter` | [`QueryContainer`](https://github.com/opensearch-project/opensearch-protobufs/blob/0.6.0/protos/schemas/common.proto#L341) | Filters for the k-NN search query. See [Filter limitations](#filter-limitations). Optional. |
+| `boost` | `float` | Boost value to increase or decrease relevance scores. Default is 1.0. Optional. |
```
```suggestion
| `boost` | `float` | A boost value to increase or decrease relevance scores. Default is 1.0. Optional. |
```

kolchfa-aws commented on _api-reference/grpc-apis/knn.md:
```diff
@@ -0,0 +1,148 @@
+| `max_distance` | `float` | The maximum physical distance in vector space for a neighbor to be considered a hit. Optional. |
+| `filter` | [`QueryContainer`](https://github.com/opensearch-project/opensearch-protobufs/blob/0.6.0/protos/schemas/common.proto#L341) | Filters for the k-NN search query. See [Filter limitations](#filter-limitations). Optional. |
+| `boost` | `float` | Boost value to increase or decrease relevance scores. Default is 1.0. Optional. |
+| `underscore_name` | `string` | Query name for query tagging (JSON name: `_name`). Optional. |
```
```suggestion
| `underscore_name` | `string` | A query name for query tagging (JSON key: `_name`). Optional. |
```

kolchfa-aws commented on _api-reference/grpc-apis/knn.md:
```diff
@@ -0,0 +1,148 @@
+| `filter` | [`QueryContainer`](https://github.com/opensearch-project/opensearch-protobufs/blob/0.6.0/protos/schemas/common.proto#L341) | Filters for the k-NN search query. See [Filter limitations](#filter-limitations). Optional. |
+| `boost` | `float` | Boost value to increase or decrease relevance scores. Default is 1.0. Optional. |
+| `underscore_name` | `string` | Query name for query tagging (JSON name: `_name`). Optional. |
+| `method_parameters` | [`ObjectMap`](https://github.com/opensearch-project/opensearch-protobufs/blob/0.6.0/protos/schemas/common.proto#L76) | Algorithm-specific parameters (`ef_search`, `nprobes`). Optional. |
```
```suggestion
| `method_parameters` | [`ObjectMap`](https://github.com/opensearch-project/opensearch-protobufs/blob/0.6.0/protos/schemas/common.proto#L76) | Algorithm-specific parameters (for example, `ef_search` or `nprobes`). Optional. |
```

kolchfa-aws commented on _api-reference/grpc-apis/knn.md:
```diff
@@ -0,0 +1,148 @@
+| `boost` | `float` | Boost value to increase or decrease relevance scores. Default is 1.0. Optional. |
+| `underscore_name` | `string` | Query name for query tagging (JSON name: `_name`). Optional. |
+| `method_parameters` | [`ObjectMap`](https://github.com/opensearch-project/opensearch-protobufs/blob/0.6.0/protos/schemas/common.proto#L76) | Algorithm-specific parameters (`ef_search`, `nprobes`). Optional. |
+| `rescore` | [`KnnQueryRescore`](https://github.com/opensearch-project/opensearch-protobufs/blob/0.6.0/protos/schemas/common.proto#L1107) | Rescoring configuration for improved accuracy. Available in versions later than 2.17. Optional. |
```
```suggestion
| `rescore` | [`KnnQueryRescore`](https://github.com/opensearch-project/opensearch-protobufs/blob/0.6.0/protos/schemas/common.proto#L1107) | A rescoring configuration for improved accuracy. Available in versions later than 2.17. Optional. |
```

kolchfa-aws commented on _api-reference/grpc-apis/knn.md:
```diff
@@ -0,0 +1,148 @@
+
+## Java gRPC client example
+
+Here's a basic example of how to use the gRPC k-NN API (actual implementation depends on your gRPC client setup):
```
```suggestion
The following is a basic example of using the gRPC k-NN API (the actual implementation depends on your gRPC client setup):
```

kolchfa-aws commented on _api-reference/grpc-apis/knn.md:
```diff
@@ -0,0 +1,148 @@
+
+## Response fields
+
+k-NN search requests return the same [`SearchResponse`](https://github.com/opensearch-project/opensearch-protobufs/blob/0.6.0/protos/schemas/search.proto#L317) structure as regular search operations. For detailed information about response fields, see [Search (gRPC) response fields]({{site.url}}{{site.baseurl}}/api-reference/grpc-apis/search/#response-fields).
```
```suggestion
k-NN search requests return the same [`SearchResponse`](https://github.com/opensearch-project/opensearch-protobufs/blob/0.6.0/protos/schemas/search.proto#L317) structure as regular search operations. For information about response fields, see [Search (gRPC) response fields]({{site.url}}{{site.baseurl}}/api-reference/grpc-apis/search/#response-fields).
```

kolchfa-aws commented on _api-reference/grpc-apis/search.md:
```diff
@@ -21,162 +21,175 @@ To submit gRPC requests, you must have a set of protobufs on the client side. Fo
+| `highlight` | [`Highlight`](https://github.com/opensearch-project/opensearch-protobufs/blob/0.6.0/protos/schemas/common.proto#L516)                     | Highlights matched terms in the result snippets. |
+| `track_total_hits` | [`TrackHits`](https://github.com/opensearch-project/opensearch-protobufs/blob/0.6.0/protos/schemas/search.proto#L309)                     | Whether to return the total hit count. |
+| `indices_boost` | `repeated` [`NumberMap`](https://github.com/opensearch-project/opensearch-protobufs/blob/0.6.0/protos/schemas/search.proto#L747)          | Per-index boost multipliers in the `<index>: <boost>` format. |
+| `docvalue_fields` | `repeated` [`FieldAndFormat`](https://github.com/opensearch-project/opensearch-protobufs/blob/0.6.0/protos/schemas/common.proto#L893)     | The fields returned using doc values, optionally formatted. |
```
```suggestion
| `docvalue_fields` | `repeated` [`FieldAndFormat`](https://github.com/opensearch-project/opensearch-protobufs/blob/0.6.0/protos/schemas/common.proto#L893)     | The fields returned using doc values. Optionally, this field can be formatted for readability. |
```

kolchfa-aws commented on _api-reference/grpc-apis/search.md:
```diff
@@ -21,162 +21,175 @@ To submit gRPC requests, you must have a set of protobufs on the client side. Fo
+| `prefilter_field` | `string` | The field to use for prefiltering to optimize script execution. Optional. |
+| `properties` | `map<string, `[`ObjectMap`](https://github.com/opensearch-project/opensearch-protobufs/blob/0.6.0/protos/schemas/common.proto#L76)`>` | Additional properties for the derived field. Optional. |
+| `ignore_malformed` | `bool` | Whether to ignore malformed values during field computation. Optional. |
+| `format` | `string` | The format to apply to the field value (e.g., date formatting). Optional. |
```
```suggestion
| `format` | `string` | The format to apply to the field value (for example., date formatting). Optional. |
```

kolchfa-aws reviewed (commented):
Thank you for submitting a thorough documentation guide, @karenyrx! Please see my comments and let me know if you have any questions. Once you address the comments, we'll move to editorial review.

karenyrx commented on _api-reference/grpc-apis/index.md:
```diff
@@ -12,23 +12,36 @@ redirect_from:
 
-1. Add the following settings to `opensearch.yml`:
+- **Reduced latency**: Binary protocol buffers eliminate JSON parsing overhead
+- **Better throughput**: More efficient network utilization for high-frequency queries
```
Changed to `Higher throughput`

karenyrx commented on _api-reference/grpc-apis/index.md:
```diff
@@ -50,29 +63,86 @@ To enable the gRPC plugin (`transport-grpc`) in OpenSearch:
+{% include copy.html %}
+
+### Notes
+- For duration-based settings (e.g., `max_connection_age`), you can use units such as `ms` (milliseconds), `s` (seconds), `m` (minutes).
```
Removed the entire `## Notes` section

karenyrx commented:
Thanks @kolchfa-aws! Addressed comments, and made one more fix: https://github.com/opensearch-project/documentation-website/pull/10651/commits/9d768f6c362cbeb403b5d907e73ed99d6e04ce5b Could you take another look when you have a chance?

kolchfa-aws commented on _query-dsl/specialized/k-nn/index.md:
```diff
@@ -10,7 +10,16 @@ redirect_from:
+
+k-NN queries can be executed using two transport protocols:
+
+- **HTTP/REST API**: The standard approach documented in this section
```
```suggestion
- **HTTP/REST API**: The standard approach documented in this section.
```

karenyrx commented on _api-reference/grpc-apis/index.md:
```diff
@@ -12,23 +12,33 @@ redirect_from:
 
-The OpenSearch gRPC plugin provides an alternative, high-performance transport layer using [gRPC](https://grpc.io/) for communication with OpenSearch. It uses protocol buffers over gRPC for lower overhead and faster serialization. This reduces overhead, speeds up serialization, and improves request-side latency, based on initial benchmarking results.
+Starting with OpenSearch version 3.2, the gRPC [Bulk API]({{site.url}}{{site.baseurl}}/api-reference/grpc-apis/bulk/) and [k-NN search queries]({{site.url}}{{site.baseurl}}/api-reference/grpc-apis/knn/) are generally available. These use [protobuf version 0.6.0](https://github.com/opensearch-project/opensearch-protobufs/releases/tag/0.6.0). However, expect updates to the protobuf structure as the ...
+{: .note}
```
The protobufs would in generally be auto-converted from the API specification, so I don't see adding protobufs to be the major challenge in the future, but rather the implementation on the server side. Is opensearch-protobufs the right place to host this issue, or core?

karenyrx replied:
But I agree it would be highly encouraged and super helpful if the community could chip in to help implementing server-side GRPC query types/other endpoints. I would propose we: 

1. Create issues in core with the `help-needed` label on what APIs or protobuf fields are open to help  on being implemented on server side (which already have the protobufs generated)
2. Continue to use https://github.com/opensearch-project/OpenSearch/issues/16787 in the public docs here, as it's the "main"/meta GRPC issue, where people can feel free to leave general suggestions on what they would like to see in GRPC 
3. Enhance the README in the opensearch-protobufs repo to encourage users to: 
     a) file issues to report their protobuf-related feedback; and 
     b) have preliminary protobufs generated for certain APIs, which they wish to help/see implemented on the server side 

What do you think?

kolchfa-aws commented on _api-reference/grpc-apis/index.md:
```diff
@@ -12,23 +12,33 @@ redirect_from:
+- **Smaller payload sizes**: Binary encoding reduces network traffic.
 
-1. Add the following settings to `opensearch.yml`:
+## Enabling gRPC APIs
```
This section is not displaying correctly when building locally. To fix, add a new line on line 42, move lines 42-53 all the way to the left (no margin), and remove `1.` on line 55.

karenyrx replied:
good catch, updated per your suggestions

kolchfa-aws reviewed (approved):
Thank you, @karenyrx! I fixed the formatting. LGTM.

karenyrx commented on _api-reference/grpc-apis/bulk.md:
```diff
@@ -439,7 +434,7 @@ Each `ResponseItem` corresponds to a single operation in the request. It contain
 ## Java gRPC client example
 
-The following example shows a Java client-side program that submits a sample bulk gRPC request and then checks whether there were any errors in the bulk response:
+The following example shows a Java client-side program that submits an example bulk gRPC request and then checks whether there were any errors in the bulk response:
```
(Sorry accidentally missed this comment) 
There is a section at the top of this page: 
```
## Prerequisite
To submit gRPC requests, you must have a set of protobufs on the client side. For ways to obtain the protobufs, see [Using gRPC APIs]({{site.url}}{{site.baseurl}}/api-reference/grpc-apis/index/#using-grpc-apis).
```
which links to the part on how the protobuf clients can be used:
```
To submit gRPC requests, you must have a set of protobufs on the client side. You can obtain the protobufs in the following ways:

- **Raw protobufs**: Download the raw protobuf schema from the [OpenSearch Protobufs GitHub repository (v0.6.0)](https://github.com/opensearch-project/opensearch-protobufs/releases/tag/0.6.0). You can then generate client-side code using the protocol buffer compilers for the [supported languages](https://grpc.io/docs/languages/).
- **Java client-side programs only**: Download the `opensearch-protobufs` jar from the [Maven Central repository](https://repo1.maven.org/maven2/org/opensearch/protobufs/0.6.0).
```
Does this seem sufficient?

karenyrx commented on _api-reference/grpc-apis/knn.md:
```diff
@@ -0,0 +1,146 @@
+
+The gRPC k-NN API has limited support for the `filter` clause compared to the HTTP API:
+
+- **Supported filters**: Basic [`term`]({{site.url}}{{site.baseurl}}/api-reference/grpc-apis/search/#term-query) and [`terms`]({{site.url}}{{site.baseurl}}/api-reference/grpc-apis/search/#terms-query) queries
```
Good call, updated!

natebower commented on _api-reference/grpc-apis/index.md:
```diff
@@ -12,67 +12,110 @@ redirect_from:
+{: .label .label-green }
 
-The OpenSearch gRPC plugin provides an alternative, high-performance transport layer using [gRPC](https://grpc.io/) for communication with OpenSearch. It uses protocol buffers over gRPC for lower overhead and faster serialization. This reduces overhead, speeds up serialization, and improves request-side latency, based on initial benchmarking results.
+Starting with OpenSearch version 3.2, the gRPC [Bulk API]({{site.url}}{{site.baseurl}}/api-reference/grpc-apis/bulk/) and [k-NN search queries]({{site.url}}{{site.baseurl}}/api-reference/grpc-apis/knn/) are generally available. These use [protobuf version 0.6.0](https://github.com/opensearch-project/opensearch-protobufs/releases/tag/0.6.0). However, expect updates to the protobuf structure as the ...
```
```suggestion
Starting with OpenSearch version 3.2, the gRPC [Bulk API]({{site.url}}{{site.baseurl}}/api-reference/grpc-apis/bulk/) and [k-NN search queries]({{site.url}}{{site.baseurl}}/api-reference/grpc-apis/knn/) are generally available. These use [protobuf version 0.6.0](https://github.com/opensearch-project/opensearch-protobufs/releases/tag/0.6.0). However, expect updates to the protobuf structure as the feature matures in upcoming versions. Other gRPC search functionality remains experimental and not recommended for production use. For updates on the progress of these features or to leave feedback, see the associated [GitHub issue](https://github.com/opensearch-project/OpenSearch/issues/16787).
```

natebower commented on _api-reference/grpc-apis/index.md:
````diff
@@ -12,67 +12,110 @@ redirect_from:
-    ```
-    {% include copy.html %}
+- **Reduced latency**: Binary protocol buffers eliminate JSON parsing overhead.
+- **Higher throughput**: More efficient network utilization for high-frequency queries
````
```suggestion
- **Higher throughput**: More efficient network utilization for high-frequency queries.
```

natebower commented on _api-reference/grpc-apis/index.md:
```diff
@@ -12,67 +12,110 @@ redirect_from:
+| `grpc.netty.worker_count` | The number of Netty worker threads for the gRPC server. Controls concurrency and parallelism. | `2` | Number of processors |
+| `grpc.netty.max_concurrent_connection_calls` | The maximum number of simultaneous in-flight requests allowed per client connection. | `200` | `100` |
+| `grpc.netty.max_connection_age` | The maximum age a connection can reach before being gracefully closed. Supports time units like `ms`, `s`, `m`. See [Time units]({{site.url}}{{site.baseurl}}/api-reference/common-parameters/#time-units). | `500ms` | Not set (no limit) |
+| `grpc.netty.max_connection_idle` | The maximum duration a connection can be idle before being closed. Supports time units like `ms`, `s`, `m`. See [Time units]({{site.url}}{{site.baseurl}}/api-reference/common-parameters/#time-units). | `2m` | Not set (no limit) |
```
```suggestion
| `grpc.netty.max_connection_idle` | The maximum duration for which a connection can be idle before being closed. Supports time units like `ms`, `s`, or `m`. See [Time units]({{site.url}}{{site.baseurl}}/api-reference/common-parameters/#time-units). | `2m` | Not set (no limit) |
```

natebower commented on _api-reference/grpc-apis/index.md:
```diff
@@ -12,67 +12,110 @@ redirect_from:
+| `grpc.publish_host` | A list of hostnames or IPs published to peers for client connections. | `["thisnode.example.com"]` | Value of `grpc.host` |
+| `grpc.netty.worker_count` | The number of Netty worker threads for the gRPC server. Controls concurrency and parallelism. | `2` | Number of processors |
+| `grpc.netty.max_concurrent_connection_calls` | The maximum number of simultaneous in-flight requests allowed per client connection. | `200` | `100` |
+| `grpc.netty.max_connection_age` | The maximum age a connection can reach before being gracefully closed. Supports time units like `ms`, `s`, `m`. See [Time units]({{site.url}}{{site.baseurl}}/api-reference/common-parameters/#time-units). | `500ms` | Not set (no limit) |
```
```suggestion
| `grpc.netty.max_connection_age` | The maximum age a connection can reach before being gracefully closed. Supports time units like `ms`, `s`, or `m`. See [Time units]({{site.url}}{{site.baseurl}}/api-reference/common-parameters/#time-units). | `500ms` | Not set (no limit) |
```

natebower commented on _api-reference/grpc-apis/index.md:
```diff
@@ -12,67 +12,110 @@ redirect_from:
+| `grpc.netty.max_connection_age` | The maximum age a connection can reach before being gracefully closed. Supports time units like `ms`, `s`, `m`. See [Time units]({{site.url}}{{site.baseurl}}/api-reference/common-parameters/#time-units). | `500ms` | Not set (no limit) |
+| `grpc.netty.max_connection_idle` | The maximum duration a connection can be idle before being closed. Supports time units like `ms`, `s`, `m`. See [Time units]({{site.url}}{{site.baseurl}}/api-reference/common-parameters/#time-units). | `2m` | Not set (no limit) |
+| `grpc.netty.keepalive_timeout` | The amount of time to wait for `keepalive` ping acknowledgment before closing the connection. Supports [time units]({{site.url}}{{site.baseurl}}/api-reference/common-parameters/#time-units). | `1s` | Not set |
+| `grpc.netty.max_msg_size` | The maximum inbound message size for gRPC requests. Supports units like `b`, `kb`, `mb`, `gb`. See [Supported units]({{site.url}}{{site.baseurl}}/api-reference/units/). | `10mb` or `10485760` | `10mb` |
```
```suggestion
| `grpc.netty.max_msg_size` | The maximum inbound message size for gRPC requests. Supports units like `b`, `kb`, `mb`, or `gb`. See [Supported units]({{site.url}}{{site.baseurl}}/api-reference/units/). | `10mb` or `10485760` | `10mb` |
```

natebower commented on _api-reference/grpc-apis/knn.md:
```diff
@@ -0,0 +1,143 @@
+{: .label .label-purple }
+
+
+The gRPC k-NN API is generally available starting with OpenSearch 3.2. However, expect updates to the protobuf structure as the feature matures in the next releases.
```
```suggestion
The gRPC k-NN API is generally available starting with OpenSearch 3.2. However, expect updates to the protobuf structure as the feature matures in upcoming versions.
```

natebower commented on _api-reference/grpc-apis/knn.md:
```diff
@@ -0,0 +1,143 @@
+| `field` | `string` | The vector field against which to run the search query. Required. |
+| `vector` | `repeated float` | A query vector. Must have the same number of dimensions as the vector field. Optional. |
+| `k` | `int32` | The number of nearest neighbors to return as top hits. Optional. |
+| `min_score` | `float` | The minimum similarity score for a neighbor to be considered a hit. Optional. |
```
```suggestion
| `min_score` | `float` | The minimum similarity score required for a neighbor to be considered a hit. Optional. |
```

natebower commented on _api-reference/grpc-apis/knn.md:
```diff
@@ -0,0 +1,143 @@
+| `vector` | `repeated float` | A query vector. Must have the same number of dimensions as the vector field. Optional. |
+| `k` | `int32` | The number of nearest neighbors to return as top hits. Optional. |
+| `min_score` | `float` | The minimum similarity score for a neighbor to be considered a hit. Optional. |
+| `max_distance` | `float` | The maximum physical distance in vector space for a neighbor to be considered a hit. Optional. |
```
```suggestion
| `max_distance` | `float` | The maximum physical distance in vector space required for a neighbor to be considered a hit. Optional. |
```

natebower commented on _api-reference/grpc-apis/knn.md:
```diff
@@ -0,0 +1,143 @@
+| `min_score` | `float` | The minimum similarity score for a neighbor to be considered a hit. Optional. |
+| `max_distance` | `float` | The maximum physical distance in vector space for a neighbor to be considered a hit. Optional. |
+| `filter` | [`QueryContainer`](https://github.com/opensearch-project/opensearch-protobufs/blob/0.6.0/protos/schemas/common.proto#L341) | Filters for the k-NN search query. See [Filter limitations](#filter-limitations). Optional. |
+| `boost` | `float` | A boost value to increase or decrease relevance scores. Default is 1.0. Optional. |
```
```suggestion
| `boost` | `float` | A boost value used to increase or decrease relevance scores. Default is 1.0. Optional. |
```

natebower commented on _api-reference/grpc-apis/knn.md:
```diff
@@ -0,0 +1,143 @@
+
+The gRPC k-NN API has limited support for the `filter` clause compared to the HTTP API. For the current list of supported query types in gRPC, see the [Search API QueryContainer documentation]({{site.url}}{{site.baseurl}}/api-reference/grpc-apis/search/#querycontainer-fields) and [Supported queries]({{site.url}}{{site.baseurl}}/api-reference/grpc-apis/search/#supported-queries).
+
+For complex filtering requirements, consider using the HTTP k-NN API, simplifying your filter logic, or waiting for the next release of k-NN gRPC.
```
```suggestion
For complex filtering requirements, consider using the HTTP k-NN API, simplifying your filter logic, or waiting for the next version of k-NN gRPC.
```

natebower commented on _api-reference/grpc-apis/knn.md:
```diff
@@ -0,0 +1,143 @@
+
+## Next steps
+
+- Learn more about [vector search in OpenSearch]({{site.url}}{{site.baseurl}}/search-plugins/knn/index/)
```
```suggestion
- Learn more about [vector search in OpenSearch]({{site.url}}{{site.baseurl}}/search-plugins/knn/index/).
```

natebower commented on _api-reference/grpc-apis/knn.md:
```diff
@@ -0,0 +1,143 @@
+## Next steps
+
+- Learn more about [vector search in OpenSearch]({{site.url}}{{site.baseurl}}/search-plugins/knn/index/)
+- Explore [k-NN index settings]({{site.url}}{{site.baseurl}}/search-plugins/knn/knn-index/)
```
```suggestion
- Explore [k-NN index settings]({{site.url}}{{site.baseurl}}/search-plugins/knn/knn-index/).
```

natebower commented on _api-reference/grpc-apis/knn.md:
```diff
@@ -0,0 +1,143 @@
+
+- Learn more about [vector search in OpenSearch]({{site.url}}{{site.baseurl}}/search-plugins/knn/index/)
+- Explore [k-NN index settings]({{site.url}}{{site.baseurl}}/search-plugins/knn/knn-index/)
+- Review [performance tuning for k-NN]({{site.url}}{{site.baseurl}}/search-plugins/knn/performance-tuning/)
```
```suggestion
- Review [performance tuning for k-NN]({{site.url}}{{site.baseurl}}/search-plugins/knn/performance-tuning/).
```

natebower commented on _api-reference/grpc-apis/knn.md:
```diff
@@ -0,0 +1,143 @@
+- Learn more about [vector search in OpenSearch]({{site.url}}{{site.baseurl}}/search-plugins/knn/index/)
+- Explore [k-NN index settings]({{site.url}}{{site.baseurl}}/search-plugins/knn/knn-index/)
+- Review [performance tuning for k-NN]({{site.url}}{{site.baseurl}}/search-plugins/knn/performance-tuning/)
+- Read about [gRPC configuration]({{site.url}}{{site.baseurl}}/api-reference/grpc-apis/index/#enabling-grpc-apis)
```
```suggestion
- Read about [gRPC configuration]({{site.url}}{{site.baseurl}}/api-reference/grpc-apis/index/#enabling-grpc-apis).
```

natebower commented on _api-reference/grpc-apis/search.md:
```diff
@@ -70,113 +70,126 @@ The [`SearchRequest`](https://github.com/opensearch-project/opensearch-protobufs
+| `prefilter_field` | `string` | The field to use for prefiltering to optimize script execution. Optional. |
+| `properties` | `map<string, `[`ObjectMap`](https://github.com/opensearch-project/opensearch-protobufs/blob/0.6.0/protos/schemas/common.proto#L76)`>` | Additional properties for the derived field. Optional. |
+| `ignore_malformed` | `bool` | Whether to ignore malformed values during field computation. Optional. |
+| `format` | `string` | The format to apply to the field value (for example., date formatting). Optional. |
```
```suggestion
| `format` | `string` | The format to apply to the field value (for example, date formatting). Optional. |
```

natebower commented on _query-dsl/specialized/k-nn/index.md:
```diff
@@ -10,7 +10,16 @@ redirect_from:
+- **HTTP/REST API**: The standard approach documented in this section.
+- **gRPC API** (Generally available 3.2): High-performance binary protocol with the [KNN (gRPC) API]({{site.url}}{{site.baseurl}}/api-reference/grpc-apis/knn/).
+
+For high-throughput vector search applications, consider using the gRPC KNN API which provides lower latency and higher throughput compared to HTTP-based queries.
```
```suggestion
For high-throughput vector search applications, consider using the gRPC KNN API, which provides lower latency and higher throughput compared to HTTP-based queries.
```

natebower reviewed (approved):
Editorial review
